### PR TITLE
openxr: Disable replay w/ unknown extensions

### PR DESCRIPTION
**Pull request by MarkY-LunarG**
Move some replay options into the base class for sharing among all the paths.  Update the OpenXR replay options to be read during replay execution.
Remove unknown extensions if the replay option is enabled.

### CODE REVIEW

**Comment by MarkY-LunarG** [comment]
Should resolve #1722

File: framework/decode/openxr_replay_consumer_base.cpp [hunk 1]
```diff
@@ -554,20 +569,26 @@ void OpenXrReplayConsumerBase::Process_xrCreateApiLayerInstance(
...
+        }
+        last_next_struct = const_cast<XrBaseInStructure*>(next_struct);
+        next_struct      = next_struct->next;
     }
-    init_android.next = reinterpret_cast<const void*>(next_struct);
-
-    XrInstanceCreateInfo new_create_info = *create_info;
-    new_create_info.next                 = &init_android;
-    create_info                          = &new_create_info;
+    init_android.next         = reinterpret_cast<const void*>(next_struct);
```
Comment (jzulauf-lunarg) [inline framework/decode/openxr_replay_consumer_base.cpp:584]: this is just setting init_android.next to nullptr, but we already updated it by using it as the first last_next_struct

File: framework/decode/replay_options.h [hunk 2]
```diff
@@ -41,25 +41,32 @@ struct ScreenshotRange
...
-    uint32_t    windowed_height{ 0 };
-    bool        force_windowed_origin{ false };
-    int32_t     window_topleft_x{ 0 };
-    int32_t     window_topleft_y{ 0 };
-    int32_t     override_gpu_index{ -1 };
-    std::string capture_filename;
-    bool        enable_print_block_info{ false };
-    int64_t     block_index_from{ -1 };
-    int64_t     block_index_to{ -1 };
+    bool                         enable_validation_layer{ false };
```
Comment (jzulauf-lunarg) [inline framework/decode/replay_options.h:44]: If were going to change anything that isn't OXR specific, that should be in a separate commit before the OXR specific one, s.t. we can put it on dev.

File: framework/decode/vulkan_replay_consumer_base.cpp [hunk 3]
```diff
@@ -5696,7 +5696,7 @@ VkResult VulkanReplayConsumerBase::OverrideCreateShaderModule(
            (pShaderModule != nullptr) && !pShaderModule->IsNull());
 
     auto original_info = pCreateInfo->GetPointer();
-    if (original_result < 0 || options_.replace_dir.empty())
+    if (original_result < 0 || options_.replace_shader_dir.empty())
```
Comment (jzulauf-lunarg) [inline framework/decode/vulkan_replay_consumer_base.cpp:5699]: see other comment re: non OXR changes

File: tools/replay/android_main.cpp [hunk 4]
```diff
@@ -156,16 +156,18 @@ void android_main(struct android_app* app)
                                                      replay_options.preload_measurement_range,
                                                      measurement_file_name);
 
-                vulkan_replay_consumer.SetFatalErrorHandler([](const char* message) { throw std::runtime_error(message); });
+                vulkan_replay_consumer.SetFatalErrorHandler(
```
Comment (jzulauf-lunarg) [inline tools/replay/android_main.cpp:159]: Accidental change?

File: tools/tool_settings.h [hunk 5]
```diff
@@ -927,8 +928,24 @@ static void GetReplayOptions(gfxrecon::decode::ReplayOptions&      options,
...
+
+    if (arg_parser.IsOptionSet(kRemoveUnsupportedOption))
+    {
+        options.remove_unsupported_features = true;
+    }
+
     IsForceWindowed(options, arg_parser);
     SetWindowOrigin(options, arg_parser);
+
+    options.screenshot_ranges      = GetScreenshotRanges(arg_parser);
```
Comment (jzulauf-lunarg) [inline tools/tool_settings.h:945]: OXR currently has no screenshot capabilities. (don't know how we'd add without implementing a compositor)

**Review by jzulauf-lunarg — CHANGES_REQUESTED** [review]
Critical: the `while (next_struct != nullptr)` loop has a couple logical errors

Stylistic: at least 3 places where an Algorithms Library approach would be cleaner/more modern

Maintainability: OXR and non-OXR changes mix in a single commit.  Not sure why the vulkan and dx options changes are needed.  If they are, should be in a separate commit

File: framework/decode/openxr_feature_util.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    }
+
+    return result;
+}
+
+bool IsSupportedApiLayer(const std::vector<XrApiLayerProperties>& properties, const char* layer)
+{
+    assert(layer != nullptr);
+
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply framework/decode/openxr_feature_util.cpp:68]: Sure, although this was a direct copy of the Vulkan code but this is cleaner.

File: framework/decode/openxr_replay_consumer_base.cpp [hunk 7]
```diff
@@ -554,20 +569,26 @@ void OpenXrReplayConsumerBase::Process_xrCreateApiLayerInstance(
...
+    const XrBaseInStructure* next_struct      = reinterpret_cast<const XrBaseInStructure*>(create_info->next);
+    while (next_struct != nullptr)
     {
-        next_struct = next_struct->next;
+        // Skip any old Android create info structs
+        if (next_struct->type == XR_TYPE_INSTANCE_CREATE_INFO_ANDROID_KHR)
+        {
+            last_next_struct->next = next_struct->next;
+        }
+        last_next_struct = const_cast<XrBaseInStructure*>(next_struct);
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply framework/decode/openxr_replay_consumer_base.cpp:581]: Good catch.

File: framework/decode/openxr_replay_consumer_base.cpp (hunk 1 above)
Comment (MarkY-LunarG) [reply framework/decode/openxr_replay_consumer_base.cpp:584]: Not if there is no Android create info.  But I'll see what I can do to clean it up.

File: framework/decode/openxr_feature_util.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    return false;
+}
+
+void RemoveUnsupportedExtensions(const std::vector<XrExtensionProperties>& properties,
+                                 std::vector<const char*>*                 extensions)
+{
+    assert(extensions != nullptr);
+
+    auto extensionIter = extensions->begin();
+    while (extensionIter != extensions->end())
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply framework/decode/openxr_feature_util.cpp:128]: This does not look like it would save us any code.  It would be roughly the same as what's here and this code was pulled from the Vulkan-side but re-purposed for OpenXR.  Plus we would have to re-define the IsSupportedExtension behavior above to work better with a non-iterator approach, and that function is being used beyond this internal case.  I think it's best to leave it as is for now.

File: framework/decode/dx_replay_options.h [hunk 9]
```diff
@@ -56,13 +56,7 @@ struct DxReplayOptions : public ReplayOptions
     bool                 override_object_names{ false };
     bool                 enable_dump_resources{ false };
     DumpResourcesTarget  dump_resources_target{};
-
-    util::ScreenshotFormat       screenshot_format{ util::ScreenshotFormat::kBmp };
```
[the remark this answers is not in the document]
Comment (MarkY-LunarG) [reply framework/decode/dx_replay_options.h:60]: I'm trying to simplify the options so that items are shared.  We can just move this into dev earlier if we want with the Vulkan-specific changes.

File: framework/decode/replay_options.h (hunk 2 above)
Comment (MarkY-LunarG) [reply framework/decode/replay_options.h:44]: Now in a separate commit.

File: framework/decode/vulkan_replay_consumer_base.cpp (hunk 3 above)
Comment (MarkY-LunarG) [reply framework/decode/vulkan_replay_consumer_base.cpp:5699]: Now in a separate commit.

File: tools/tool_settings.h (hunk 5 above)
Comment (MarkY-LunarG) [reply tools/tool_settings.h]: Yes, but DX and Vulkan both  shared this code.  But it is now outside of the OpenXR-specific changes.

File: framework/decode/openxr_replay_consumer_base.cpp (hunk 1 above)
Comment (MarkY-LunarG) [reply framework/decode/openxr_replay_consumer_base.cpp]: It is now cleaned up.

File: tools/replay/android_main.cpp (hunk 4 above)
Comment (MarkY-LunarG) [reply tools/replay/android_main.cpp]: Not there any more.

**Review by jzulauf-lunarg — COMMENTED** [review]
LGTM

**Review by jzulauf-lunarg — APPROVED** [review]
